api/debug.go

```diff
@@ -68,3 +113,189 @@ func TraceStructLog(ctx *rpc.CallContext, db state.StateDB, header *types.Header
 	}
 	return tracer.StructLogs(), nil
 }
+
+// TraceCall lets you trace a given eth_call. It collects the structured logs
+// created during the execution of EVM if the given transaction was added on
+// top of the provided block and returns them as a JSON object.
+// If no transaction index is specified, the trace will be conducted on the state
+// after executing the specified block. However, if a transaction index is provided,
+// the trace will be conducted on the state after executing the specified transaction
+// within the specified block.
+func (s *PrivateDebugAPI) TraceCall(ctx *rpc.CallContext, args TransactionArgs, blockNrOrHash *vm.BlockNumberOrHash, config *TraceCallConfig) (interface{}, error) {
+	bNrOrHash := vm.BlockNumberOrHashWithNumber(rpc.PendingBlockNumber)
+	if blockNrOrHash != nil {
+		bNrOrHash = *blockNrOrHash
+	}
+	var (
+		statedb state.StateDB
+		header *types.Header
+		chaincfg *params.ChainConfig
+		evmFn evmfunc
+	)
+
+	if config != nil  && config.TxIndex != nil {
+		var err error
+		statedb, header, evmFn, chaincfg, err = s.stateAtTransaction(ctx, bNrOrHash, int(*config.TxIndex))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		err := s.evmmgr.View(bNrOrHash, args.from, ctx, func(h *types.Header, sdb state.StateDB, getEVM func (state.StateDB, *vm.Config, common.Address, *big.Int) *vm.EVM, cfg *params.ChainConfig) error{
+			statedb = sdb
```

Comment:
I would not expect this to work, at least not reliably.

The `state.StateDB` object accesses the database through a `storage.Transaction` that only exists for the life of the View callback function. Once the view function returns, I would expect the storage.Transaction object to start returning errors, and if it doesn't do so immediately it's going to as soon as the database is done cleaning up the transaction.

While it definitely complicates code structure, any interactions with the statedb object need to happen within the view functions.